content/blog/2022/2022-10-25-getting-started-with-spring-security.md

```diff
@@ -0,0 +1,1011 @@
+`UserAuthenticationErrorHandler` and `UserForbiddenErrorHandler`.
+
+With this configuration, the postman response for the REST endpoint looks like this:
+
+Success Response:
+
+{{% image alt="settings" src="images/posts/getting-started-with-spring-security/postman01.JPG" %}}
+
+Unauthorized Response:
+
+{{% image alt="settings" src="images/posts/getting-started-with-spring-security/postman-unauth.JPG" %}}
+
+Forbidden Response:
+
+{{% image alt="settings" src="images/posts/getting-started-with-spring-security/postman-forbidden.JPG" %}}
+
+#### 3. Additional endpoints secured by default
+Once spring security is configured for a request matcher, additional endpoints added get secured by default.
+For instance, let's add an endpoint to the `BookController` class
+````java
+@GetMapping("/library/books/all")
+    public ResponseEntity<List<BookDto>> getAllBooks() {
+        return ResponseEntity.ok().body(bookService.getAllBooks());
+    }
+````
+For this endpoint to be called successfully, we need to provide basic auth credentials.
+
+Error response when no credentials passed:
+
+{{% image alt="settings" src="images/posts/getting-started-with-spring-security/postman-nocreds.JPG" %}}
+
+Success response:
+
+{{% image alt="settings" src="images/posts/getting-started-with-spring-security/postman-creds.JPG" %}}
+
+#### 4. Unsecure Specific Endpoints
+We can specify a list of endpoints that need to be excluded from the security configuration.
+To achieve this, let's first add another endpoint to our `BookController` class and add the below configuration:
+````java
+@GetMapping("/library/info")
+    public ResponseEntity<LibraryInfo> getInfo() {
+        return ResponseEntity.ok().body(bookService.getLibraryInfo());
+    }
+````
+
+````java
+@Bean
+    public WebSecurityCustomizer webSecurityCustomizer() {
+        return (web) -> web.ignoring().antMatchers("/library/info");
+    }
+````
+Now, we should be able to hit the endpoint from postman without passing credentials:
+
+{{% image alt="settings" src="images/posts/getting-started-with-spring-security/unsecure.JPG" %}}
+
+#### 5. Add Custom Filters
+Spring provides security by executing a sequence of filters in a chain. In cases where we need to add additional checks to the request
+before it reaches the controller, Spring Security provides us with the below methods that help us add a custom filter at the desired position in the chain.
+- **addFilterBefore(Filter filter, Class<? extends Filter> beforeFilter)**: This method lets us add the custom filter before the specified filter in the chain.
+- **addFilterAfter(Filter filter, Class<? extends Filter> afterFilter)**: This method lets us add the custom filter after the specified filter in the chain.
+- **addFilterAt(Filter filter, Class<? extends Filter> atFilter)**: This method lets us add the custom filter at the specified filter in the chain.
```

Comment:
What does it mean? Is the old filter overridden at this location? If not, in which order will they be executed?